source/common/stats/symbol_table_impl.h

```diff
@@ -0,0 +1,122 @@
+#pragma once
+
+#include <algorithm>
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include "envoy/common/exception.h"
+#include "envoy/stats/symbol_table.h"
+
+#include "common/common/assert.h"
+
+#include "absl/strings/str_join.h"
+#include "absl/strings/str_split.h"
+
+namespace Envoy {
+namespace Stats {
+
+/**
+ * Underlying SymbolTableImpl implementation which assists with per-symbol reference counting.
+ *
+ * The underlying Symbol / SymbolVec data structures are not meant for public consumption. One side
+ * effect of the non-monotonically-increasing symbol counter is that if a string is encoded, the
+ * resulting stat is destroyed, and then that same string is re-encoded, it may or may not encode to
+ * the same underlying symbol.
+ */
+class SymbolTableImpl : public SymbolTable {
+public:
+  StatNamePtr encode(const std::string& name) override;
```

Comment:
can you try making this return std::unique_ptr<StatNameImpl>?  That would help avoid the dynamic_cast in your tests. I'm not 100% sure if that's a valid covariant return, but it is worth a shot.

I think the dynamic_casts you have are OK if you need to have them, but let's see if we can avoid it if it's easy.

Reply:
I think that covariant returns won't work here because of a weird cross-dependency between SymbolTableImpl and StatNameImpl. StatNameImpl expects a SymbolTableImpl& to store internally, so if we declare a covariant return it's unclear which of the two impls to write first, since they both would depend on each other. I'm comfortable with the `dynamic_cast<>` if it avoids this problem.